tests/lib/rules/no-constant-condition.js

```diff
@@ -103,6 +108,12 @@ ruleTester.run("no-constant-condition", rule, {
         { code: "if(abc==='str' || true){}", errors: [{ message: "Unexpected constant condition.", type: "IfStatement" }] },
         { code: "if(abc==='str' || true || def ==='str'){}", errors: [{ message: "Unexpected constant condition.", type: "IfStatement" }] },
         { code: "if(false || true){}", errors: [{ message: "Unexpected constant condition.", type: "IfStatement" }] },
-        { code: "if(typeof abc==='str' || true){}", errors: [{ message: "Unexpected constant condition.", type: "IfStatement" }] }
+        { code: "if(typeof abc==='str' || true){}", errors: [{ message: "Unexpected constant condition.", type: "IfStatement" }] },
+
+        { code: "function* foo(){while(true){} yield 'foo';}", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Unexpected constant condition.", type: "WhileStatement" }] },
+        { code: "function* foo(){while(true){if (true) {yield 'foo';}}}", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Unexpected constant condition.", type: "IfStatement" }] },
+        { code: "function* foo(){if (true) {yield 'foo';}}", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Unexpected constant condition.", type: "IfStatement" }] },
+        { code: "function* foo(){while(true){yield 'foo';}while(true){}}", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Unexpected constant condition.", type: "WhileStatement" }] },
+        { code: "var a = function* foo(){while(true){} yield 'foo';}", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Unexpected constant condition.", type: "WhileStatement" }] }
```

Comment:
A few more test cases to consider:

```js
// This should be valid because the `yield` would pause both loops.
while (true) {
  while (true) {
    yield;
  }
}
```

```js
// This should be invalid because the `yield` would not pause the loop.
while (true) {
  function* foo() {
    yield;
  }
}
```

```js
// This should be invalid because the `yield` does not get executed on every loop iteration.
function* foo() {
  for (let foo = yield; ;) {}
}
```

```js
// This should be valid because the `yield` gets executed on every loop iteration.
function* foo() {
  for (; yield; ) {}
}
```

```js
// This should be valid because the `yield` gets executed on every loop iteration.
function* foo() {
  for (; ; yield) {}
}
```

```js
// This should be valid because the second `yield` would pause the loop.
function* foo() {
  while (true) {
    function* foo() {
      yield;
    }
    yield;
  }
}
```

```js
// invalid for the outer loop, valid for the inner loop
function foo() {
  while (true) {
    function* bar() {
      while (true) {
        yield;
      }
    }
  }
}
```